packages/compiler/src/ml_parser/ast.ts

```diff
@@ -10,60 +10,69 @@ import {AstPath} from '../ast_path';
 import {AST as I18nAST} from '../i18n/i18n_ast';
 import {ParseSourceSpan} from '../parse_util';
 
-export interface Node {
+interface BaseNode {
   sourceSpan: ParseSourceSpan;
   visit(visitor: Visitor, context: any): any;
 }
 
-export class Text implements Node {
+export type Node = Attribute|CDATA|Comment|DocType|Element|Text;
+// Expansion|ExpansionCase -- not used
+
+export class Text implements BaseNode {
   constructor(public value: string, public sourceSpan: ParseSourceSpan, public i18n?: I18nAST) {}
   visit(visitor: Visitor, context: any): any { return visitor.visitText(this, context); }
+  readonly type = 'text';
```

Comment:
I thought you meant to expose the node classes directly, e.g.,

```ts
// angular-html-parser
export { Text, CDATA, ... } from 'path/to/ast.ts'
```

so that you can easily identify the node via `node instanceof Text` and similar. Any reason why add the `type` field instead? I think it'd be better to not modify the angular internal thing if possible. What do you think?

Reply:
Adding this property turns `Node` into a tagged union, which, for example, helped find this bug:
```diff
-  visitCdata(text: Text, context: any): any;
+  visitCdata(text: CDATA, context: any): any;
```
Also string properties like this are kind of standard for AST formats.

But reverting this part is not a problem as it can be done via declaration merging on Prettier's side. Should I revert it?

Reply:
Make sense.